### PR TITLE
Schema versioning

### PR DESCRIPTION
**Pull request by joshmoore**
The code was updated to expect `tile_shape` to be a dict rather than a list:

```
{"x": 2048, "y": 2048}    # not: [2048, 2048]
```

but the _schema_ was updated to accept either. This was required since it was only possible to validate against the current schema. By encoding the version of each schema in its path (and storing *all* schemas in the same directory), it's possible to choose the correct validator at runtime based on the version in the json document.

 - [x] rename existing schemas with a version in their path
 - [x] update tests to use the new locations
 - [x] add helpers to remove leaky path strings
 - [x] add docs to new classes
 - [ ] ~check docs for deprecated/removed usage~ I've forgotten what I wanted to do here.
 - [ ] ~minimize the number of version strings in the code base~ see https://github.com/spacetx/starfish/pull/1278#issuecomment-491302107

Requires ongoing cleanup:
 - [ ] add old schemas
 - [ ] add tests for old schemas
 - [ ] verify `$Id` of all schemas have the appropriate tag
 - [ ] review tests for use of `0.0.0`

Potential follow-ups:
 - [ ] clarify starfish's use of `$id`. see [docs](https://json-schema.org/understanding-json-schema/structuring.html#the-id-property)
 - (potentially useful) force the use of a given schema version
 - (valuable) always validate the tree

Test plan:
 - travis should be green

### CODE REVIEW

File: starfish/core/spacetx_format/test_coordinates.py [hunk 1]
```diff
@@ -4,8 +4,9 @@
 from .util import SpaceTxValidator
 
 package_name = "starfish"
+current_version = "0.1.0"
```
Comment (ttung) [inline starfish/core/spacetx_format/test_coordinates.py]: https://github.com/spacetx/slicedimage/pull/95

File: starfish/core/spacetx_format/test_experiment.py [hunk 2]
```diff
@@ -3,12 +3,12 @@
...
 from pkg_resources import resource_filename
 
-from .util import SpaceTxValidator
+from .util import ExperimentValidator
 from .validate_sptx import validate
 
 package_name = "starfish"
-experiment_schema_path = resource_filename(package_name, "spacetx_format/schema/experiment.json")
-validator = SpaceTxValidator(experiment_schema_path)
+current_version = "5.0.0"
```
Comment (ttung) [inline starfish/core/spacetx_format/test_experiment.py]: starfish.core.experiment.experiment.Experiment.version.CURRENT_VERSION

File: starfish/core/spacetx_format/test_fuzz.py [hunk 3]
```diff
@@ -1,17 +1,11 @@
...
+from .util import CodebookValidator, ExperimentValidator, Fuzzer
 
 package_name = "starfish"
-codebook_schema_path = resource_filename(
-    package_name, "spacetx_format/schema/codebook/codebook.json")
-experiment_schema_path = resource_filename(
-    package_name, "spacetx_format/schema/experiment.json")
-codebook_validator = SpaceTxValidator(codebook_schema_path)
-experiment_validator = SpaceTxValidator(experiment_schema_path)
+codebook_validator = CodebookValidator(version="0.0.0")
```
Comment (ttung) [inline starfish/core/spacetx_format/test_fuzz.py]: Should this be based on the current version?

File: starfish/spacetx_format/schema/experiment_4.0.0.json [hunk 4]
```diff
@@ -0,0 +1,51 @@
+{
+  "$id": "https://github.com/spacetx/starfish/blob/0.0.22/validate_sptx/schema/experiment.json",
```
Comment (ttung) [inline starfish/spacetx_format/schema/experiment_4.0.0.json]: I thought you were going to point this at the commit prior to 5.0.0

File: starfish/core/spacetx_format/test_coordinates.py (hunk 1 above)
Comment (joshmoore) [reply starfish/core/spacetx_format/test_coordinates.py]: You see keeping the coordinates.json schema in sync with the slicedimage version?

**Comment by joshmoore** [comment]
To get rid of more hard-coded versions, I'd likely have to make the `starfish/core/spacetx_format` package the owner, and import that elsewhere:

```
(sf01) [jamoore@idr1-slot2 starfish]$ git grep -E '[0-9][.][0-9][.][0-9]' -- starfish/{core,spacetx_format}/**/*.py
starfish/core/codebook/_format.py:CURRENT_VERSION = Version("0.0.0")
starfish/core/codebook/_format.py:MIN_SUPPORTED_VERSION = Version("0.0.0")
starfish/core/codebook/_format.py:MAX_SUPPORTED_VERSION = Version("0.0.0")
starfish/core/experiment/experiment.py:        if version < Version("5.0.0"):
starfish/core/experiment/version.py:CURRENT_VERSION = Version("5.0.0")
starfish/core/experiment/version.py:MIN_SUPPORTED_VERSION = Version("4.0.0")
starfish/core/experiment/version.py:MAX_SUPPORTED_VERSION = Version("5.0.0")
starfish/core/spacetx_format/test_codebook.py:current_version = "0.0.0"
starfish/core/spacetx_format/test_coordinates.py:current_version = "0.1.0"
starfish/core/spacetx_format/test_field_of_view.py:current_version = "0.1.0"
starfish/core/spacetx_format/test_fov_manifest.py:current_version = "0.1.0"
starfish/core/spacetx_format/test_fuzz.py:. . . . . .      0.0.0
starfish/core/spacetx_format/test_fuzz.py:        "version": "0.0.0",
starfish/core/spacetx_format/test_fuzz.py:. . . . . .      0.0.0
starfish/core/spacetx_format/util.py:        # Note: we are using 5.0.0 here as the first known file. It does *not* need to
starfish/core/spacetx_format/util.py:            package_name, "spacetx_format/schema/experiment_5.0.0.json")
starfish/core/spacetx_format/util.py:LatestCodebookValidator = lambda: CodebookValidator(version="0.0.0")
starfish/core/spacetx_format/util.py:LatestExperimentValidator = lambda: ExperimentValidator(version="5.0.0")
starfish/core/spacetx_format/util.py:LatestFOVValidator = lambda: FOVValidator(version="0.1.0")
starfish/core/spacetx_format/util.py:LatestManifestValidator = lambda: ManifestValidator(version="0.1.0")
```

File: starfish/spacetx_format/schema/experiment_4.0.0.json (hunk 4 above)
Comment (joshmoore) [reply starfish/spacetx_format/schema/experiment_4.0.0.json]: i.e. just 0.0.21 or the actual commit? Sorry. It paged out.

**Comment by joshmoore** [comment]
The infrastructure for multiple versions should now all more or less be in place. Remaining open TODOs are around choosing historical versions based on the git history and getting them into this code base (and optionally writing example files against them that are tested). I'd suggest getting this in with a minimum of changes and then then at least future versions can use the infrastructure.

File: starfish/core/spacetx_format/test_coordinates.py (hunk 1 above)
Comment (ttung) [reply starfish/core/spacetx_format/test_coordinates.py]: That's the plane.

File: starfish/spacetx_format/schema/experiment_4.0.0.json (hunk 4 above)
Comment (ttung) [reply starfish/spacetx_format/schema/experiment_4.0.0.json]: Any commit that references the 4.0.0 schema should be ok.

File: starfish/core/spacetx_format/test_coordinates.py (hunk 1 above)
Comment (joshmoore) [reply starfish/core/spacetx_format/test_coordinates.py]: :+1: Suggestion in f1c6078f7ec4f6f234b6bd52e3788389ecf49e9c

File: starfish/spacetx_format/schema/experiment_4.0.0.json (hunk 4 above)
Comment (joshmoore) [reply starfish/spacetx_format/schema/experiment_4.0.0.json]: Ok. The version variable was updated in 16ecf10da06dbf3e4bae7ad67e4f43419383e4ed, so using its parent: d933c08a6987da26433714f23cf331e067087bf6

Comment (joshmoore) [reply starfish/spacetx_format/schema/experiment_4.0.0.json]: (Note: that commit is somewhere between 0.0.26 and 0.0.27)

File: starfish/core/spacetx_format/test_coordinates.py [hunk 5]
```diff
@@ -1,11 +1,15 @@
 import pytest
 from pkg_resources import resource_filename
+from slicedimage import VERSIONS
```
Comment (ttung) [inline starfish/core/spacetx_format/test_coordinates.py]: I think you need #1334 for this. :(

Looks like travis runs are broken here.

Comment (joshmoore) [reply starfish/core/spacetx_format/test_coordinates.py]: :+1: I pip installed locally assuming that'd be the thing to do :wink:  Merged master and pushed.

**Comment by joshmoore** [comment]
Gree

**Pull request closed without merging**